crates/rustc_codegen_spirv/src/abi.rs

```diff
@@ -843,5 +843,43 @@ fn trans_intrinsic_type<'tcx>(
                 Err(ErrorReported)
             }
         }
+        IntrinsicType::Matrix => {
+            let (elem_type, n) =
+                if let Abi::Vector { ref element, count } = ty.field(cx, 0).layout.abi {
+                    match &element.value {
+                        Primitive::F32 => Some(SpirvType::Float(32)),
+                        Primitive::F64 => Some(SpirvType::Float(64)),
+                        Primitive::Int(integer, signed) => {
+                            Some(SpirvType::Integer(integer.size().bits() as u32, *signed))
+                        }
+                        Primitive::Pointer => None,
+                    }
+                    .map(|elem_type| (elem_type, count as u32))
+                } else {
+                    None
+                }
+                .ok_or_else(|| {
+                    cx.tcx
+                        .sess
+                        .err("#[spirv(matrix)] type must have a vector with scalar element type");
+                    ErrorReported
+                })?;
+
+            let m = ty.fields.count() as u32;
+
+            let elem_spirv = elem_type.def(span, cx);
+
+            let vector = SpirvType::Vector {
+                element: elem_spirv,
+                count: n as u32,
+            }
+            .def(span, cx);
+
+            Ok(SpirvType::Matrix {
+                element: vector,
+                count: m as u32,
+            }
+            .def(span, cx))
+        }
```

Comment:
You don't have to (and shouldn't) do this digging into field types for this, and should use the standard type translation tools on the fields instead. Also, this is missing quite a bit of validation (e.g. if fields are different types, or there are no fields - IIRC your code ICEs if there's no fields) - there should be tests for each of the three kinds of error. Something like this should work instead:

```rust
IntrinsicType::Matrix => {
    let field_types = (0..ty.fields.count())
        .map(|i| trans_type_impl(cx, span, ty.field(cx, i), false))
        .collect::<Vec<_>>();
    if field_types.is_empty() {
        cx.tcx
            .sess
            .err("#[spirv(matrix)] type must have at least one field");
        return Err(ErrorReported);
    }
    let elem_type = field_types[0];
    if !field_types.iter().all(|&ty| ty == elem_type) {
        cx.tcx
            .sess
            .err("#[spirv(matrix)] type fields must all be the same type");
        return Err(ErrorReported);
    }
    match cx.lookup_type(elem_type) {
        SpirvType::Vector { .. } => (),
        ty => {
            cx.tcx
                .sess
                .struct_err("#[spirv(matrix)] type fields must all be vectors")
                .note(&format!("field type is {}", ty.debug(elem_type, cx)))
                .emit();
            return Err(ErrorReported);
        }
    }

    Ok(SpirvType::Matrix {
        element: elem_type,
        count: field_types.len() as u32,
    }
    .def(span, cx))
}
```

Reply:
Thanks.
I updated the code and change Matrix length validation since `OpTypeMatrix` requires a length of at least 2.
https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpTypeMatrix